cmd/ctl/pkg/upgrade/migrate/migrator.go

```diff
@@ -0,0 +1,264 @@
+	return nil
+}
+
+func (m *Migrator) ensureCRDStorageVersionEquals(ctx context.Context, vers string, names []string) (bool, []*apiext.CustomResourceDefinition, error) {
+	var crds []*apiext.CustomResourceDefinition
+	for _, crdName := range names {
+		crd := &apiext.CustomResourceDefinition{}
+		if err := m.Client.Get(ctx, client.ObjectKey{Name: crdName}, crd); err != nil {
+			return false, nil, err
+		}
+
+		// Discover the storage version
+		storageVersion := storageVersionForCRD(crd)
+
+		if storageVersion != vers {
+			fmt.Fprintf(m.Out, "CustomResourceDefinition object %q has storage version set to %q. You MUST upgrade to cert-manager v1.0-v1.6 before migrating resources for v1.7.\n", crdName, storageVersion)
+			return false, nil, nil
+		}
+
+		crds = append(crds, crd)
+	}
+
+	return true, crds, nil
+}
+
+func (m *Migrator) discoverCRDsRequiringMigration(ctx context.Context, desiredStorageVersion string, names []string) ([]*apiext.CustomResourceDefinition, error) {
+	var requireMigration []*apiext.CustomResourceDefinition
+	for _, name := range names {
+		crd := &apiext.CustomResourceDefinition{}
+		if err := m.Client.Get(ctx, client.ObjectKey{Name: name}, crd); err != nil {
+			return nil, err
+		}
+		// If no versions are stored, there's nothing to migrate.
+		if len(crd.Status.StoredVersions) == 0 {
+			continue
+		}
+		// If more than one entry exists in `storedVersions` OR if the only element in there is not
+		// the desired version, perform a migration.
+		if len(crd.Status.StoredVersions) > 1 || crd.Status.StoredVersions[0] != desiredStorageVersion {
+			requireMigration = append(requireMigration, crd)
+		}
+	}
+	return requireMigration, nil
+}
+
+func (m *Migrator) migrateResourcesForCRD(ctx context.Context, crd *apiext.CustomResourceDefinition) error {
+	startTime := time.Now()
+	fmt.Fprintf(m.Out, "Migrating %q objects in group %q - this may take a while (started at %s)...\n", crd.Spec.Names.Kind, crd.Spec.Group, startTime.Format(time.Stamp))
+	list := &unstructured.UnstructuredList{}
+	list.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: "v1",
+		Kind:    crd.Spec.Names.ListKind,
+	})
+	if err := m.Client.List(ctx, list); err != nil {
+		return err
+	}
+	fmt.Fprintf(m.Out, " %d resources to migrate\n", len(list.Items))
+	for _, obj := range list.Items {
+		// retry on any kind of error to handle cases where e.g. the network connection to the apiserver fails
+		if err := retry.OnError(wait.Backoff{
+			Duration: time.Second, // wait 1s between attempts
+			Steps:    3,           // allow up to 3 attempts per object
+		}, func(err error) bool {
+			// Retry on any errors that are not otherwise skipped/ignored
+			return handleUpdateErr(err) != nil
+		}, func() error { return m.Client.Update(ctx, &obj) }); handleUpdateErr(err) != nil {
+			return err
+		}
+	}
```

Comment:
Can you call `Update` on an `UnstructuredList`? If so perhaps that'd be more reliable. I.e. just send the whole list to the API server and let it sort it out.

Reply:
I also recall this comment from Jordan in https://kubernetes.slack.com/archives/C0EG7JC6T/p1641313521453000?thread_ts=1641312183.447800&cid=C0EG7JC6T:
> 2xx response from a CR patch request means the content is at the configured storage version. once all existing CRs have been patched, you're good to clean up the CRD status 

Patch rather than Update? Does it matter?

A Patch operator would  also be less likely to fail due to optimistic locking, wouldn't it?

Reply:
OTOH I suppose if the Update fails with an optimistic locking error, we can assume that the CR has been stored using the storage version.

Reply:
> Can you call Update on an UnstructuredList? If so perhaps that'd be more reliable. I.e. just send the whole list to the API server and let it sort it out.

Nope, updates have to be called individually by the client. There's no way to make the apiserver perform a batch UPDATE operation (as it's not possible to perform transactions in the apiserver, which means there's no way for the apiserver to ensure all succeed or all fail)

> OTOH I suppose if the Update fails with an optimistic locking error, we can assume that the CR has been stored using the storage version.

Correct, and we purposely absorb/skip over `IsConflict` and `IsNotFound` errors in `handleUpdateError` 😄